Fix for query exec when abort is called before exec

Aklakan opened the pull request:
GitHub issue resolved #2394

Pull request Description: Cancel signal is moved to class level and is set to true when abort is called.



----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages. (test case and fix are separate commits)
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

namedgraph commented:
Reminds me of this issue I had but I never got around to isolating it...
https://www.mail-archive.com/users@jena.apache.org/msg20564.html

Aklakan commented:
I added another test case for concurrent abort which should (occasionally) hang prior to this patch - the reason it should hang is that sometimes abort happens to be called before the execution starts - this scenario is also covered by the non-concurrent test case.

if it doesn't hang, then maybe the test's task count per cpu needs to be increased - feedback is welcome.

afs commented:
Improving this is good to have.

To try my understanding out, I wrote some tests. Maybe I'm misunderstanding the intention.

Whole file: [TestCancelQueryAFS.java.txt](https://github.com/apache/jena/files/14980629/TestCancelQueryAFS.java.txt)

It is what happens for CONSTRUCT et al, and whether abort before exec does cancel when there is no iteration.

`test_cancel_select_1()` does not cancel. It does not touch the iterator.
`test_cancel_select_2()` does cancel.

In the tests file there are tests for CONSTRUCT and ASK. I haven't figured out why these don't cancel - I would have thought the iterator was run.

There is also a copy of test_Cancel_API_5 and variations to show I am runnign the PR code :smile:

```java
    @Test
    public void test_cancel_select_1() {
        cancellationTest("SELECT * {}", exec->exec.select());
    }

    @Test
    public void test_cancel_select_2() {
        cancellationTest("SELECT * {}", exec->exec.select().hasNext());
    }

. . .

    static void cancellationTest(String queryString, Consumer<QueryExec> execAction) {
        DatasetGraph dsg = DatasetGraphFactory.createTxnMem();
        dsg.add(SSE.parseQuad("(_ :s :p :o)"));

        try(QueryExec aExec = QueryExec.dataset(dsg).query(queryString).build()) {
            aExec.abort();
            assertThrows(QueryCancelledException.class, ()-> execAction.accept(aExec));
        }
    }
```

Aklakan commented:
test_cancel_select_1 does not raise the exception because the iterator is not accessed - the cancel flag is set though.
IMO QueryCancelledException should only be thrown on iterator access - if it were raised directly on `select()` then in general one would need two exception handlers - one on the iterator and one on the query exec level. The try-with-resources block closes resources of a query exec whose iterator has not been touched - so there is shouldn't be a resource leak.

I have so far only looked at the select case (not construct, describe, etc) - should I look into this and extend the PR or will you work on it?

Aklakan commented:
> So I'd say that the correct behavior

Hm, "correctness" is the wrong wording - its an API contract question. The query exec methods that do not return an iterator are expected to fail right away (because they'll internally touch the iterator). However, if an iterator is returned I am uncertain whether the creation of the iterator should already fail due to cancellation - or whether the place to check for cancellation of an iterator is solely on next() (IIRC hasNext returns true even if cancelled, but picking the item up with next() will raise the exception.).

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/QueryExecDataset.java:
```diff
@@ -89,6 +89,7 @@ public class QueryExecDataset implements QueryExec
     private long                     timeout2         = TIMEOUT_UNSET;
     private final AlarmClock         alarmClock       = AlarmClock.get();
     private long                     queryStartTime   = -1; // Unset
+    private AtomicBoolean            cancelSignal     = new AtomicBoolean(false);
```
Because this has moved to the class, it does not need to be passed as an argument to the constructors of the inner classes.  

Because the classes have members of the same name, I'm getting a compiler warning.

Aklakan replied:
Removed the superfluous ctor argument from the inner classes.

afs commented:
Thanks for pointing out that failing at `select()` is a visible change.

I'm personally in favour of consistency (my perceived idea of consistency!) which is fail ASAP. The `QueryExec` has been created and the abort operation is on that.

```
    qexec.abort();
    qexec.select();
```

`select()` does already get questions asked so why is it quick and then the query is slow.

`construct()` isn't failing nor is `ask()`. `describe()` is.

`ask()` calls `.next()` once and the comment says something about timeouts.

`.hasNext()` is often the place where the real work has to happen because to know whether there is a "next" , the engine has to do the work!

Suggestion: merge this PR as a step forward (a couple of warnings have appeared - `cancelSignal` is now a member and being hidden by the inner class timeout) addressing #2394 and do more in another issue/PR.

Aklakan commented:
> I'm personally in favour of consistency (my perceived idea of consistency!) which is fail ASAP. 

Right, I agree that fail fast is typically easier to debug.

I noticed that `execute()` was only used for select and describe, whereas `startQueryIterator()` is the method actually common to all operations - so I moved the abort check there.

I also extended your iterator-based test cases to make two assertions whether abort happens before or after obtaining the iterator.

The test cases now work according to our understanding.

Aklakan commented on jena-arq/src/test/java/org/apache/jena/sparql/api/TestCancelQueryAFS.java:
```diff
@@ -0,0 +1,155 @@
+package org.apache.jena.sparql.api;
+
+/*
+ *  Copyright (c) Telicent Ltd.
```
License header likely needs adjustment (this is the only file with a Telicent header)

Aklakan commented on jena-arq/src/test/java/org/apache/jena/sparql/api/TestCancelQueryAFS.java:
```diff
@@ -0,0 +1,155 @@
+import org.apache.jena.vocabulary.RDF;
+import org.junit.Test;
+
+public class TestCancelQueryAFS {
```
Merge into TestQueryExecutionCancel?

afs replied:
Yes. And drop the tests that are copies (`test_Cancel_API_5`).

Aklakan replied:
Methods are now merged into TestQueryExecutionCancel. Duplicated tests are removed.

afs commented:
Looks good.

I'll do some clearup - like not call a test class "... AFS". I've been looking at the Java memory model recently and there a volatile which could at least be an `AtomicReference` and maybe use fences, not full volatile semantics. It won't make an observable difference in this situation but it least the pattern will be recorded - not going as far as `VarHandles` which look quite complicated and fragile for general code.

Aklakan commented:
I removed the class TestQueryAFS and moved the methods to TestQueryExecutionCancel.

Are you going to adapt the code from volatile to AtomicReference?

afs commented:
Excellent.

Let's merge this and do any refinement as and when time permits.
